app/src/main/java/com/fivegmag/a5gmsmediasessionhandler/MediaSessionHandlerMessengerService.kt

```diff
@@ -302,5 +323,177 @@ class MediaSessionHandlerMessengerService() : Service() {
+            Toast.LENGTH_LONG
+        ).show()
+
+        // call m5 report consumption to AF - TS26.512 Clause 4.7.4
+        val consumptionReportingApi: ConsumptionReportingApi? = clientsSessionData[sendingUid]?.consumptionReportingApi
+        if(consumptionReportingApi == null)
+        {
+            Log.i(TAG, "[ConsumptionReporting] consumptionReportingApi is 【NULL】")
+            return
+        }
+
+        val provisisioningSessionId: String = "2";
+        val call: Call<ResponseBody>? = consumptionReportingApi.postConsumptionReporting(provisisioningSessionId, dataReporting?.reportingClientId);
+
+        call?.enqueue(object : retrofit2.Callback<ResponseBody> {
+            override fun onResponse(
+                call: Call<ResponseBody?>,
+                response: Response<ResponseBody?>
+            ) {
+                Log.i(TAG, "[ConsumptionReporting] resp from AF:" + response.body()?.string())
+            }
+
+            override fun onFailure(call: Call<ResponseBody?>, t: Throwable) {
+                Log.i(TAG, "[ConsumptionReporting] onFailure")
+                call.cancel()
+            }
+        })
+    }
+    
+    private fun createRetrofitForConsumpReport(clientId: Int) {
+        try {
+            var consumpReportUrl: String = "";
+            if(clientsSessionData[clientId]?.serviceAccessInformation!!.clientConsumptionReportingConfiguration.serverAddresses.isNotEmpty())
+            {
+                consumpReportUrl = clientsSessionData[clientId]?.serviceAccessInformation!!.clientConsumptionReportingConfiguration.serverAddresses[0];
+            }
+            Log.i(TAG, "[ConsumptionReporting] createRetrofitForConsumpReport URL: $consumpReportUrl.")
+
+            if (consumpReportUrl != "") {
+                val retrofit = retrofitBuilder
+                    .baseUrl(consumpReportUrl)
+                    .build()
+
+                clientsSessionData[clientId]?.consumptionReportingApi = retrofit.create(ConsumptionReportingApi::class.java)
+            }
+        } catch (e: Exception) {
+            Log.e(TAG, "[ConsumptionReporting] onException of createRetrofitForConsumpReport")
+        }
+    }
+
+    fun startConsumptionReportTimer(clientId: Int) {
+        val timer = Timer()
+        clientsSessionData[clientId]?.serviceAccessInformationRequestTimer = timer
+
+        if(clientsSessionData[clientId]?.serviceAccessInformation?.clientConsumptionReportingConfiguration?.reportingInterval != 0U) {
+            var periodSec: UInt? =
+                clientsSessionData[clientId]?.serviceAccessInformation?.clientConsumptionReportingConfiguration!!.reportingInterval
+            Log.i(TAG, "[ConsumptionReporting] startConsumptionReportTimer periodSec: $periodSec.")
+
+            timer.schedule(
+                object : TimerTask() {
+                    override fun run() {
+                        clientsSessionData[clientId]?.isConsumptionReportByTimer = true
```

Comment:
Consumption reports shall be send periodically: 

_Reports are sent at periodic intervals determined by the reporting interval attribute of the consumption reporting configuration specified in Table 4.2.3-2._

With this logic it seems like the next report is send only if a message from the MediaStreamHandler was dispatched and not at the predefined interval.

Reply:
I feel that my understanding is not quite the same as yours, and I only found:
- In 4.7.2.3: The Media Session Handler shall periodically check for updated Service Access Information
- In 4.7.4: Upon determining the need to report ongoing 5GMS consumption at periodic intervals determined by the clientConsumptionReportingConfiguration.reportingInterval property.

As I understood, the report is send only if a message from the MediaStreamHandler was dispatched, and the conditions 4.7.4 occur including check the timer at the predefined interval. Am I misunderstood?

Reply:
The consumption reports are sent as defined by `reportingInterval`. With each report the `duration` property is increased, see also here: https://github.com/5G-MAG/rt-5gms-media-stream-handler/issues/52